copernicus_download/download.py

```diff
@@ -1,18 +1,148 @@
+        Starting point for crawler
+    download_directory : str
+        Directory to save files to (default: current directory)
+    username : str
+        Username required for authentication (default: no authentication)
+    password : str
+        Password required for authentication (default: no authentication)
+    include : glob pattern
+        Download only matching files (default: download all files)
+    exclude : glob pattern
+        Do not download matching files (default: download all files)
+    download_jobs : int
+        Number of concurrent jobs used for downloading files (default: 10)
+    crawler_args : dict
+        Keyword arguments to pass to itsybitsy.crawl
+    """
+    logger.debug("crawling %s", base_url)
+
+    if crawler_args is None:
+        crawler_args = {}
+
+    with requests.Session() as session:
+        if username and password:
+            session.auth = (username, password)
+
+        if crawler_args.get("session") is None:
+            crawler_args["session"] = session
+
+        crawler = itsybitsy.crawl(base_url, **crawler_args)
+        base_url_normalized = next(crawler)
+        base_path = urlparse.urlparse(base_url_normalized).path
+
+        if isinstance(include, str):
+            include = [include]
+        if isinstance(exclude, str):
+            exclude = [exclude]
+
+        futures = set()
+        with concurrent.futures.ThreadPoolExecutor(max_workers=download_jobs) as executor:
+            for url in crawler:
+                logger.debug("> found link: %s" % url)
+                url_parts = urlparse.urlparse(url)
+                file_path = url_parts.path
+                if not file_path.startswith(base_path):
+                    warnings.warn("File {} does not match base path {} - skipping".format(file_path, base_path))
+                    continue
+
+                target_localpath = os.path.normpath(file_path[len(base_path):])
+                target_fullpath = os.path.join(download_directory, target_localpath)
+
+                if include and not any(fnmatch.fnmatch(target_localpath, pattern) for pattern in include):
+                    logger.debug(">> skipping due to include pattern")
+                    continue
+                if exclude and any(fnmatch.fnmatch(target_localpath, pattern) for pattern in exclude):
+                    logger.debug(">> skipping due to exclude pattern")
+                    continue
+
+                try:
+                    os.makedirs(os.path.dirname(target_fullpath))
+                except OSError:
+                    pass
+
+                logger.debug(">> downloading")
+                future = executor.submit(_download_file, url, target_fullpath, session)
+                futures.add(future)
+
+            if futures:
+                for future in concurrent.futures.as_completed(futures):
+                    yield future.result()
+
+
+def download_data(url, username, password, download_dir='.', include='*.zip'):
```

Comment:
What else is interface-breaking apart from `include` replacing `data_ext`?

Reply:
`download_data` is an iterator now. I also didn't want to keep the `data_ext` name around, since the functionality is different.